### PR TITLE
Allow resources to contain unexported fields

### PR DESCRIPTION
**Pull request by scothis**
Using unexported fields in Kubernetes resources is not common, but does happen. Previously, these fields would cause cmp.Diff to panic. Now, we ignore the content of unexported fields when computing a diff. As the diff is for logging and test assertions this is relatively safe.

Semantic equality is used for detecting when a managed resource needs to be updated. Custom equality func can be defined separately as needed.

Refs #484

### CODE REVIEW

File: internal/resources/resource_with_unexported_fields.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+func (r *TestResourceUnexportedFields) CopyUnexportedFields() {
+	r.Status.unexportedFields = r.Spec.unexportedFields
+}
+
+// +kubebuilder:object:generate=true
+type TestResourceUnexportedFieldsSpec struct {
+	Fields           map[string]string `json:"fields,omitempty"`
+	unexportedFields map[string]string
```
Comment (neowulf) [inline internal/resources/resource_with_unexported_fields.go:77]: it wasn't clear to me whether the cmp IgnoreUnexported option works recursively or for the top level struct only. IMHO, it'd be useful to have a nested struct with unexported fields as part of this spec.

Comment (neowulf) [reply internal/resources/resource_with_unexported_fields.go:77]: I see you have implemeted your own opt. Does it handle nested structs with unexported fields?

Comment (scothis) [reply internal/resources/resource_with_unexported_fields.go:77]: Within cmp, each option is applied to every path discovered while walking the object tree. If the path is filtered no fields under it are processed.

File: testing/config.go [hunk 2]
```diff
@@ -427,6 +427,7 @@ var (
 		return str != "" && !strings.HasPrefix(str, "Status")
 	}, cmp.Ignore())
 
+	// Deprecated: use reconcilers.IgnoreAllUnexported instead
```
Comment (mamachanko) [inline testing/config.go:430]: **[praise]:** Thanks for this! Will schedule removal in next+1 version. https://github.com/search?q=SafeDeployDiff&type=code suggests that it's not used by public projects on Github at least.

**Review by mamachanko — APPROVED** [review]
⭐ → @scothis! Thank you very much for the effort. I specifically took note of how you are driving it with `TestResourceUnexportedFields`.

@neowulf would you be able to test-drive this branch with your kit?

Since `ExpectConfig` is using `IgnoreAllUnexported` to verify all CRUD events, there would no more immediate need to open up cmp options for configuration as far as https://github.com/vmware-labs/reconciler-runtime/issues/484 is concerned, or is there? Correct me if I am wrong.

**Comment by mamachanko** [comment]
**[question, non-blocking]:** @scothis when you say

> Custom equality func can be defined separately as needed.

are you meaning that there could potentially be a follow-up PR? I am asking because afaik there's currently no way to provide a custom equality func, or is there?

**Comment by scothis** [comment]
@mamachanko I was basing it on this comment https://github.com/vmware-labs/reconciler-runtime/issues/484#issuecomment-1958682986

File: reconcilers/cmp.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// IgnoreAllUnexported is a cmp.Option that ignores unexported fields in all structs
+var IgnoreAllUnexported = cmp.FilterPath(func(p cmp.Path) bool {
+	// from cmp.IgnoreUnexported with type info removed
+	sf, ok := p.Index(-1).(cmp.StructField)
+	if !ok {
+		return false
+	}
+	r, _ := utf8.DecodeRuneInString(sf.Name())
+	return !unicode.IsUpper(r)
+}, cmp.Ignore())
```
[the remark this answers is not in the document]
Comment (scothis) [reply reconcilers/cmp.go:24]: Funny, we cribbed from the same source to arrive at the identical outcome

**Comment by neowulf** [comment]
@scothis this fix works! Thank you again for the quick turnaround!

**Comment by scothis** [comment]
@neowulf cool, I'll circle back to Max's feedback tomorrow.

File: internal/resources/resource_with_unexported_fields.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+func (r *TestResourceUnexportedFieldsStatus) GetUnexportedFields() map[string]string {
+	return r.unexportedFields
+}
+
+func (r *TestResourceUnexportedFieldsStatus) SetUnexportedFields(f map[string]string) {
+	r.unexportedFields = f
+}
+
+func (r *TestResourceUnexportedFieldsStatus) AddUnexportedFields(key, value string) {
```
[the remark this answers is not in the document]
Comment (scothis) [reply internal/resources/resource_with_unexported_fields.go:152]: I typically use `Add*` when modifying a single key in a map, and `Set*` to replace the whole map. I will update to make it singular.

File: reconcilers/cmp.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+package reconcilers
+
+import (
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+// IgnoreAllUnexported is a cmp.Option that ignores unexported fields in all structs
```
[the remark this answers is not in the document]
Comment (scothis) [reply reconcilers/cmp.go:15]: done

File: reconcilers/resource_test.go [hunk 6]
```diff
@@ -1282,3 +1282,124 @@ func TestResourceReconciler(t *testing.T) {
...
+				resource,
+			},
+			Metadata: map[string]interface{}{
+				"SubReconciler": func(t *testing.T, c reconcilers.Config) reconcilers.SubReconciler[*resources.TestResourceUnexportedFields] {
+					return &reconcilers.SyncReconciler[*resources.TestResourceUnexportedFields]{
+						Sync: func(ctx context.Context, resource *resources.TestResourceUnexportedFields) error {
+							if resource.Status.Fields == nil {
+								resource.Status.Fields = map[string]string{}
+							}
+							resource.CopyUnexportedFields()
```
[the remark this answers is not in the document]
Comment (scothis) [reply reconcilers/resource_test.go]: done

File: reconcilers/child_test.go [hunk 7]
```diff
@@ -2108,3 +2108,232 @@ func TestChildReconciler_Unstructured(t *testing.T) {
...
+			ExpectEvents: []rtesting.Event{
+				rtesting.NewEvent(resource, scheme, corev1.EventTypeNormal, "Updated",
+					`Updated TestResourceUnexportedFields %q`, testName),
+			},
+			ExpectResource: resourceReady.
+				SpecDie(func(d *dies.TestResourceSpecDie) {
+					d.AddField("foo", "bar")
+					d.AddField("new", "field")
+				}).
+				DieReleasePtr(),
```
[the remark this answers is not in the document]
Comment (scothis) [reply reconcilers/child_test.go]: done

File: internal/resources/resource_with_unexported_fields.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+	if r.Spec.Fields != nil {
+		if _, ok := r.Spec.Fields["invalid"]; ok {
+			field.Invalid(field.NewPath("spec", "fields", "invalid"), r.Spec.Fields["invalid"], "")
+		}
+	}
+
+	return errs
+}
+
+func (r *TestResourceUnexportedFields) CopyUnexportedFields() {
```
[the remark this answers is not in the document]
Comment (scothis) [reply internal/resources/resource_with_unexported_fields.go]: done

File: internal/resources/resource_with_unexported_fields.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (scothis) [reply internal/resources/resource_with_unexported_fields.go]: done

**Review by neowulf — APPROVED** [review]

**Comment by mamachanko** [comment]
Thank you @scothis !

**Pull request merged**